Better document the requirements file format

maxwell-k opened the pull request:
Change the documentation for the requirements file format so that it matches the [implementation].

Before this change the documentation included an incomplete list of supported options.

This change adds the missing options and changes the order to match, so that the two locations are easier to keep in sync. After this change the list in the documentation matches `SUPPORTED_OPTIONS` in `src/pip/_internal/req/req_file.py`.

[implementation]: https://github.com/pypa/pip/blob/master/src/pip/_internal/req/req_file.py#L60

Addresses https://github.com/pypa/pip/issues/7385

chrahunt commented on docs/html/reference/pip_install.rst:
```diff
@@ -153,14 +153,19 @@ To interpret the requirements file in UTF-8 format add a comment
+  *  :ref:`-i, --index-url <install_--index-url>`
   *  :ref:`-f, --find-links <install_--find-links>`
+  *  :ref:`--extra-index-url <install_--extra-index-url>`
+  *  ``-Z, --always-unzip``
```
This is undocumented and deprecated (see [here](https://github.com/pypa/pip/blob/9ae2e6440d9fc98e91acec0a324b410fabc5452a/src/pip/_internal/cli/cmdoptions.py#L799-L806)), so we can exclude it.

maxwell-k replied:
Thanks, I noticed it is undocumented but missed the deprecation, thanks for the pointer!

maxwell-k commented:
> Just one comment, otherwise this is OK for me.

Addressed in latest push, thank you.

> The new ordering is a little odd (index options are not together, for example) but no problem fixing it in both places later.

I agree. To address this I have adjusted the order in the implementation, so that the documentation isn't so odd! I am happy to action any feedback on ordering.

> If you want, as a followup (you can make a separate issue if you are interested) it may also be possible to generate this list automatically from SUPPORTED_OPTIONS similar to what we do for command option help here.

I would like to try that and will open a separate issue and PR later.   Thank you for the suggestion.  I'll keep this PR to improving the hand written documentation.

chrahunt commented:
Something is up with Azure, we'll give it some time.
